### PR TITLE
Add defsfont integration test

### PR DESCRIPTION
**Pull request by mawe42**
As discussed in #813, we thought it might be useful to have an integration test that tests the SoundFont loader, SF parser and defsfont code all in one go. This PR adds a test utility `dump_sfont` that outputs the internal structure of a `fluid_defsfont_t` as a YAML file. It also adds a previously dumped version for the VintageDreams SoundFont (sf2 and sf3) and adds tests to make sure the current code still results in the same output.

I've also added some code to fluid_gen.c, as I wanted a way to output generator names without having to duplicate all the generator definitions. But that could probably also be done differently.

This is a request for comments on the approach as a whole, not a final version ready for merge.

### CODE REVIEW

**Comment by derselbst** [comment]
This looks very good already! Yaml is also a good choice (haven't thought of this before, I guess it was too obvious :) ). I'll probably have a closer look at this during the (extended) easter holidays.

**Comment by mawe42** [comment]
Thanks! :) And yes... YAML seemed like a really good choice, especially for C programs that need to output structured data. Simple to write but quite expessive.

I'm having trouble with the Windows CI runs... do you have a quick idea on how to specify the path to the compiled `dump_sfont` executable? I've tried: `${CMAKE_CURRENT_BINARY_DIR}/dump_sfont${CMAKE_EXECUTABLE_SUFFIX}` and `${CMAKE_BINARY_DIR}/test/dump_sfont${CMAKE_EXECUTABLE_SUFFIX}`, but that both seems to result in `D:/a/1/s/build/test/dump_sfont.exe`. But the compiled binary is in `D:\a\1\s\build\test\Release\dump_sfont.exe`.

(https://dev.azure.com/tommbrt/tommbrt/_build/results?buildId=3978&view=logs&j=abc15840-b60a-5377-a855-7d49c24083b5&t=d8190f29-a713-5c4c-e34a-d9389633ff32&l=118)

**Comment by mawe42** [comment]
> o you have a quick idea on how to specify the path to the compiled dump_sfont executable?

I think I've found a solution:
https://stackoverflow.com/a/56201564

Using a dummy generator expression (whatever that is) seems to be the recommended way of removing the per-configuration sub-directories in VS and Xcode at the moment.

**Comment by mawe42** [comment]
Hm... this is weird: the test on MinGW x86 does seem to use libinstpatch to load SoundFonts by default:
https://dev.azure.com/tommbrt/tommbrt/_build/results?buildId=3984&view=logs&j=2e75b761-8d38-561a-da0c-385616ba7153

**Comment by derselbst** [comment]
> the test on MinGW x86 does seem to use libinstpatch to load SoundFonts by default

That's an old problem that lies somehwere between `_fseeki64`, `fgetpos` and a presumably broken mingw x86 implementation of those functions, making the defsont loader fail, causing it to ask libinstpatch. Just ignore it.

**Review by derselbst — APPROVED** [review]
I'm already quite happy with this one. Just two nit-picks below. If you want, feel free to look through the coverage and see if there are any more missing test cases. But otherwise, we could also add them later on if need be.

File: test/dump_sfont.c [hunk 1]
```diff
@@ -0,0 +1,430 @@
...
+            continue;
+        }
+
+        indent();
+        dump_inst_zone(zone);
+        outdent();
+    }
+}
+
+static int inst_compare_func(void *a, void *b)
```
[the remark this answers is not in the document]
Comment (mawe42) [reply test/dump_sfont.c]: Good point, fixed.

File: test/dump_sfont.c [hunk 2]
```diff
@@ -0,0 +1,430 @@
...
+        goto EXIT;
+    }
+
+    sfont = fluid_synth_get_sfont_by_id(synth, id);
+    if (sfont == NULL)
+    {
+        goto EXIT;
+    }
+
+    if (sfont->free != fluid_defsfont_sfont_delete)
```
[the remark this answers is not in the document]
Comment (mawe42) [reply test/dump_sfont.c]: I guess it does make some sense. "Explicit is better than implicit"... oh wait... that's Python ;-)

**Comment by mawe42** [comment]
> If you want, feel free to look through the coverage and see if there are any more missing test cases. But otherwise, we could also add them later on if need be.

I think I would vote to add more test cases if and when we need them. Full code coverage is not a goal in itself, in my opinion.

If we decide that the unit tests in #826 are the way to go, then I would suggest to remove the test fonts for broken fonts from this PR again. No need to test this twice, only adds more maintenance burden in the future.

File: test/dump_sfont.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (mawe42) [reply test/dump_sfont.c]: Hm... SonarQube **really** doesn't like the cast to the non-const compare func...

[the remark this answers is not in the document]
Comment (mawe42) [reply test/dump_sfont.c]: And removing the case makes gcc unhappy. I think I will revert this again. The `fluid_list_compare_func` should really have `const void` arguments. Maybe fix that first?

[the remark this answers is not in the document]
Comment (derselbst) [reply test/dump_sfont.c:291]: Oh right. This comes from the `fluid_compare_func_t`. I missed that. Then keep it non-const for now. We'll do that separately.

**Comment by derselbst** [comment]
>  I would suggest to remove the test fonts for broken fonts from this PR again. No need to test this twice, only adds more maintenance burden in the future.

Ok for me.

**Comment by derselbst** [comment]
Once you're done you can mark this as ready. Then I would merge master back in and resolve conflicts.

**Comment by mawe42** [comment]
Ok, I've removed the broken SoundFont tests, cleaned up the commit history a little bit and rebased it against current master. So from my perspective this is finished.

**Pull request merged**